### PR TITLE
new post: What is the OAuth 2.0 Authorization Code Grant Type?

### PR DESCRIPTION
**Pull request by aaronpk**
new post: What is the OAuth 2.0 Authorization Code Grant Type?

to be published on 4/10

### CODE REVIEW

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+description: "The Authorization Code Grant Type is used by both web apps and native apps to get an access token after a user authorizes an app. This post is the first part of a series where we explore the frequently used OAuth 2.0 grant types."
+tags: [oauth]
+tweets:
+- "What is the #oauth2 Authorization Code Grant Type? Read @aaronpk’s article here:"
+- "Ever wondered exactly what the #oauth2 Authorization Code Grant Type was for? Now you can stop wondering! ;)"
+---
+
+The Authorization Code Grant Type is probably the most common of the OAuth 2.0 grant types that you'll encounter. It is used by both web apps and native apps to get an access token after a user authorizes an app.
+
+This post is the first part of a series where we explore frequently used OAuth 2.0 grant types. If you want to back up a bit and learn more about OAuth 2.0 before we dive in, check out [What the Heck is OAuth?](https://developer.okta.com/blog/2017/06/21/what-the-heck-is-oauth), also on the Okta developer blog.
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: Most links to https://developer.okta.com should be relative, so this should be `/blog/...`

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+* `response_type=code` - This tells the authorization server that the application is initiating the authorization code flow.
+* `client_id` - The public identifier for the application, obtained when the developer first registered the application.
+* `redirect_uri` - Tells the authorization server where to send the user back to after they approve the request.
+* `scope` - One or more space-separated strings indicating which permissions the application is requesting. The specific OAuth API you're using will define the scopes that it supports.
+* `state` - The application generates a random string and includes it in the request. It should then check that the same value is returned after the user authorizes the app. This is used to prevent [CSRF attacks](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29).
+
+When the user visits this URL, the authorization server will present them with a prompt asking if they would like to authorize this application's request.
+
+![OAuth Prompt](https://oauth2simplified.com/images/oauth-prompt.png)
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: You can center images using the by adding the `center-image` CSS class. I'm not sure how this works for images on another domain. We usually use something like following for images.

```
{% img blog/react-native-app-auth/ios-id-token.png alt:"ID Token on iOS" width:"500" %}{: .center-image }
```

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+We're about ready to wrap up the flow. Now that the application has the authorization code, it can use that to get an access token.
+
+The application makes a POST request to the service's token endpoint with the following parameters:
+
+* `grant_type=authorization_code` - This tells the token endpoint that the application is using the Authorization Code grant type.
+* `code=` - The application includes the authorization code it was given in the redirect.
+* `redirect_uri=` - The same redirect URI that was used when requesting the code. Some APIs don’t require this parameter, so you’ll need to double check the documentation of the particular API you’re accessing.
+* `client_id=` - The application's client ID.
+* `client_secret=` - The application's client secret. This ensures that the request to get the access token is made only from the application, and not from a potential attacker that may have intercepted the authorization code.
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: The previous bullet list does not have a trailing equals on the parameters. I'd suggest removing them here too.

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+The token endpoint will verify all the parameters in the request, ensuring the code hasn't expired and that the client ID and secret match. If everything checks out, it will generate an access token and return it in the response!
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Cache-Control: no-store
+Pragma: no-cache
+
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: If you want syntax highlighting on the JSON, you can end this code block and add ```json to the one below.

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+The code exchange step ensures that an attacker isn’t able to intercept the access token, since the access token is always sent via a secure backchannel between the application and the OAuth server.
+
+## Learn More About OAuth and Okta
+
+You can learn more about OAuth 2.0 on [OAuth.com](https://www.oauth.com/), or check out any of these resources to get started building!
+* [Get Started with Spring Boot, OAuth 2.0, and Okta](https://developer.okta.com/blog/2017/03/21/spring-boot-oauth)
+* [Token Authentication in ASP.NET Core 2.0 - A Complete Guide](https://developer.okta.com/blog/2018/03/23/token-authentication-aspnetcore-complete-guide)
+* [Secure your SPA with Spring Boot and OAuth](https://developer.okta.com/blog/2017/10/27/secure-spa-spring-boot-oauth)
+Or hit up [Okta’s OIDC/OAuth 2.0 API](http://developer.okta.com/docs/api/resources/oidc) for specific information on how we support OAuth. And as always, follow us on Twitter @oktadev for more great content.
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: These links should be relative.

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md (hunk 3 above)
Comment (aaronpk) [reply _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: good catch

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md (hunk 2 above)
Comment (aaronpk) [reply _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: Thanks for catching that, I forgot this was at an external URL I was using for testing. I'm copying it into this site.

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md (hunk 4 above)
Comment (aaronpk) [reply _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: Looks like it adds a blank line when I separate them, so I'd rather not have JSON highlighting and keep them in the same code block.

File: _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md [hunk 6]
```diff
@@ -112,8 +112,8 @@ The code exchange step ensures that an attacker isn’t able to intercept the ac
...
-* [Get Started with Spring Boot, OAuth 2.0, and Okta](https://developer.okta.com/blog/2017/03/21/spring-boot-oauth)
-* [Token Authentication in ASP.NET Core 2.0 - A Complete Guide](https://developer.okta.com/blog/2018/03/23/token-authentication-aspnetcore-complete-guide)
-* [Secure your SPA with Spring Boot and OAuth](https://developer.okta.com/blog/2017/10/27/secure-spa-spring-boot-oauth)
-Or hit up [Okta’s OIDC/OAuth 2.0 API](http://developer.okta.com/docs/api/resources/oidc) for specific information on how we support OAuth. And as always, follow us on Twitter @oktadev for more great content.
+* [Get Started with Spring Boot, OAuth 2.0, and Okta](/blog/2017/03/21/spring-boot-oauth)
+* [Token Authentication in ASP.NET Core 2.0 - A Complete Guide](/blog/2018/03/23/token-authentication-aspnetcore-complete-guide)
+* [Secure your SPA with Spring Boot and OAuth](/blog/2017/10/27/secure-spa-spring-boot-oauth)
+Or hit up [Okta’s OIDC/OAuth 2.0 API](/docs/api/resources/oidc) for specific information on how we support OAuth. And as always, follow us on Twitter @oktadev for more great content.
```
Comment (mraible) [inline _source/_posts/2018-04-10-oauth-authorization-code-grant-type.md]: I would add a link around "follow us on Twitter @oktadev"

**Pull request merged**